### ISSUE TITLE
Stateless function components cannot be given refs

### ISSUE THREAD

**Issue by mattdell** [vitkon/form-container#28]
## Expected Behavior
No error when using a stateless component

## Actual Behavior
```
Warning: Stateless function components cannot be given refs. Attempts to access this ref will fail.
```

## Steps to Reproduce the Problem

  1. Use `bindInput` on a stateless component

## Specifications

  - Version: 1.0.4
  - Platform: Chrome

What value is the `ref` prop providing?

**Comment by vitkon** [comment]
`ref` is enabling HTML5 attributes validation, e.g. `<input name="foo" required/>`

**Comment by mattdell** [comment]
Hmm... so what to do about the errors given when using a stateless function component as they don't support using a `ref`?

**Comment by vitkon** [comment]
I'll look into it

**Comment by mattdell** [comment]
Seems this can be fixed with a simple check for `Component.prototype.render;`

https://stackoverflow.com/questions/41488713/react-how-to-determine-if-component-is-stateless-functional

Full example: 
```
        bindInput = (name: keyof T): IBoundInput => {
            const inputMethods: IBoundInput = {
                name,
                value: this.getValue(name),
                onChange: this.bindToChangeEvent,
                onFocus: this.bindToFocusEvent,
                onBlur: this.bindToBlurEvent,
            };

            if (React.Component.prototype.render) {
                inputMethods.ref = (input: any) => {
                    if (!this.state.inputs[name]) {
                        this.setState({
                            inputs: {
                                ...this.state.inputs,
                                [name]: input
                            }
                        });
                    }
                }
            }

            return inputMethods;
        };
```

I'll raise a PR

**Comment by vitkon** [comment]
@mattdell how can I replicate? Seems like it's working with my stateless components

**Comment by mattdell** [comment]
It's more like this when the ref is sent to the stateless function component rather than a native input...

```
const Input = (props: any) => (
  <input {...props} />
);

const Form = () => (
  <Input {...bindInput('username')} />
);
```

**Comment by vitkon** [comment]
Yeah, I kind of figured it. Tried with rebass Input component.
It'd be good to verify that it works with Input components from major component libs (Material, Semantic, React BS, Ant Design, etc).

Let me look into it further to decide what to do with refs.
Don't want to ditch HTML5 validation, it's kind of a unique feature

**Comment by mattdell** [comment]
I'll do some investigating...

**Comment by mattdell** [comment]
So with Material has the same result using this code:

```
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { connectForm, IFormProps, } from 'form-container';
import { TextField as MaterialInput } from 'material-ui';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';

const Input = (props: any) => (
  <MaterialInput {...props} />
);

const Form = (props: IFormProps) => (
  <MuiThemeProvider>
    <Input {...props.formMethods.bindInput('username')} />
  </MuiThemeProvider>
);

const ConnectedForm = connectForm()(Form);

ReactDOM.render(
  <ConnectedForm />,
  document.getElementById('root') as HTMLElement,
);
```

```
index.js:2177 Warning: Stateless function components cannot be given refs. Attempts to access this ref will fail.

Check the render method of `Form`.
    in Input (created by Form)
    in MuiThemeProvider (created by Form)
    in Form (created by validated)
    in validated (created by FormWrapper)
    in FormWrapper
```

I see in v1 beta they expose an `inputRef` prop to the native input but I couldn't get that version to work locally. I suppose that is the way forward with most of these libraries. https://material-ui-next.com/api/input/

I suppose the question is do we need to enforce that every bound input has access to an internal ref? This could be quite limiting.

**Comment by vitkon** [comment]
limiting it is def not the way forward

**Comment by mattdell** [comment]
Agreed. Well, we can't take advantage of HTML5 validation without a reference (literally) to the native input. And we're also assuming that each bound input actually has a native input (the client could be doing weird stuff with just a div and CSS). Can we only take advantage of HTML5 validation when there is a native input, and it is exposed? I'm not sure how to determine that.

**Comment by mattdell** [comment]
Looked into it a bit more...

Is there any way of getting the reference to the input we're binding when we say `...bindInput`? Presumably not because that's simply deconstructing an object. Otherwise the best solution I can think of is having to pass it manually like so...

```
const Input = (props: any) => <input { ...props } />;

class MockComponent extends React.Component<any, any> {
  render() {
    const { formMethods: { bindInput }, form } = this.props

    return (
      <form>
        <Input {...bindInput('foo', { isStateless: true }) } />
      </form>
    )
  }
}
```

Alternatively you could `bindInput` and `bindStatefulInput`?

I'm not sure wrapping it in a stateful HOC like recompose would achieve much benefit (like HTML5 validation) other than getting rid of the warning.

**Comment by vitkon** [comment]
I like `bindStatefulInput` idea, should do the job for now

**Comment by mattdell** [comment]
Merged #31
